include/elements/2d/quadrilateral_bspline_element.tcc

```diff
@@ -63,6 +85,27 @@ inline Eigen::VectorXd
       }
     }
 
+    //! If kernel correction is needed
+    if (apply_kernel_correction) {
+      // Compute M inverse matrix
+      Eigen::Matrix<double, Tdim + 1, Tdim + 1> M =
+          Eigen::Matrix<double, Tdim + 1, Tdim + 1>::Zero();
+      for (unsigned n = 0; n < this->nconnectivity_; ++n) {
+        Eigen::Matrix<double, Tdim + 1, 1> p;
+        p << 1.0, (nodal_coordinates_.row(n).transpose() - pcoord);
+        M.noalias() += shapefn(n) * p * p.transpose();
+      }
+      const auto& M_inv = M.inverse();
+      const double C1 = M_inv(0, 0);
+      const VectorDim& C2 = M_inv.block(1, 0, Tdim, 1);
+
+      // Corrected shape function calculation
+      for (unsigned n = 0; n < this->nconnectivity_; ++n) {
+        shapefn(n) *=
+            (C1 + C2.dot(nodal_coordinates_.row(n).transpose() - pcoord));
```

Comment:
Any reference to check these lines? I can not review these lines without reading the paper.

Reply:
I will pass you privately.

Reply:
Thank you, I will read the document and check the PR again